### DISCUSSION TITLE
Monkeypatching filter() to support block scoping

### DISCUSSION THREAD

**Discussion by richard-sim** [premake/premake-core#2266] [Ideas]
After one too many dangling `filter()` calls resulting in hard-to-track-down bugs, I decided to see what I could do about patching `filter()` to bring some of the premake-6 promised land to premake5. And... it's GLORIOUS!

This patch still supports the "legacy" style filter usage, so is a drop-in fix for your own premake scripts while not breaking 3rd party libraries (or your own that you haven't ported to the block-scoped style yet).

```lua
-- Patch the filter() function to add block scoping
local monkeypatched_filter = filter
function filter(terms, scopedBlock)
    -- The orignal filter function does nothing if terms is nil, but we
    -- don't want to run scopedBlock in that case. This is a change in
    -- behaviour where previously the code that's now inside scopedBlock
    -- would have been run and affected the outer scope's configuration.
    if terms then
        monkeypatched_filter(terms)
        if scopedBlock then
            scopedBlock()
            monkeypatched_filter({})
        end
    end
end
```

Usage:
```lua
filter("kind:StaticLib", function ()
    defines { "L1a_STATIC" }
end)
filter("kind:SharedLib", function ()
    defines { "L1a_SHARED" }
end)

-- "legacy" filters still work, but ew
filter "kind:StaticLib"
    defines { "L1a_STATIC" }
filter "kind:SharedLib"
    defines { "L1a_SHARED" }
filter {}
```

**Comment by tritao** [comment]
That's pretty cool, I wonder if we could have something like this for all containers.

Seems like it's something we would be able to support without breaking compatibility, so might be worth investigating further.

And could get us closer to Premake 6 idealized syntax.

**Reply by richard-sim** [reply]
> I wonder if we could have something like this for all containers.

I considered looking into the others, but, honestly, this was born out of frustration with `filter` and I haven't experienced the same frustrations elsewhere, so passed on taking it any further at the time.

> might be worth investigating further

I'd love to see it integrated into mainline so everyone can benefit from it. I hesitated on making this a PR, as it feels like a big step (a significant usage change to probably the single most called function in premake, even if it's fully backwards compatible) and, being new to premake, I wasn't sure if that was "how we do things around here". 🙃

**Reply by novamon** [reply]
this is what I did to write a frontend over premake to extend library tracking, workspace templating and a module system to load preconfigured modules for external libraries like SDL2, Image libs, etc. The goal was to track what dependencies a library needed and properly include them and their configurations required, automatically into the child project being built where needed. Simplified making the projects a great deal.
